### PR TITLE
[Bugfix] Fix bug mysql display array type in mysql-client 8.x

### LINKED ISSUE #9202: mysql 8.0 query array type display incorrectly
### Steps to reproduce the behavior (Required)
mysql-client 5 is ok 8 is wrong
1. mysql> select [1, 2, 3] as numbers;

### Expected behavior (Required)
mysql> select [1,2,3] as numbers;
+---------+
| numbers |
+---------+
| [1,2,3] |
+---------+
1 row in set (0.11 sec)
### Real behavior (Required)
mysql> mysql> select [1, 2, 3] as numbers;
+------------------+
| numbers          |
+------------------+
| 0x5B312C322C335D |
+------------------+
### StarRocks version (Required)
 - You can get the StarRocks version by executing SQL `select current_version()`
2.4

### PR DESCRIPTION
**Pull request by Astralidea**
## What type of PR is this：
- [x] bug
- [ ] feature
- [ ] enhancement
- [ ] refactor
- [ ] others

## Which issues of this PR fixes ：

Fixes #9202

## Problem Summary(Required) ：

This is because we do not define the character encoding of the Array column at the protocol layer. In MySQL 8, it will be treated as binary, and in 5, the character encoding will be used by default.

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/catalog/PrimitiveType.java [hunk 1]
```diff
@@ -65,7 +65,7 @@ public enum PrimitiveType {
     DECIMAL128("DECIMAL128", 16, TPrimitiveType.DECIMAL128),
 
     JSON("JSON", 16, TPrimitiveType.JSON),
-
+    ARRAY("ARRAY", -1, TPrimitiveType.ARRAY),
```
Comment (imay) [inline fe/fe-core/src/main/java/com/starrocks/catalog/PrimitiveType.java:68]: Array is not primitive type, you should use TypeDesc to check if it is Array

File: fe/fe-core/src/main/java/com/starrocks/catalog/Type.java [hunk 2]
```diff
@@ -1484,7 +1484,11 @@ public int getMysqlResultSetFieldCharsetIndex() {
             case BITMAP:
                 return 33;
             default:
-                return 63;
+                if (this instanceof ArrayType) {
```
Comment (imay) [inline fe/fe-core/src/main/java/com/starrocks/catalog/Type.java:1487]: why not override this function in ArrayType?

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 3 / 3 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/ArrayType.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/Type.java|2|2|100.00%|[]|

**Comment by Astralidea** [comment]
run starrocks_admit_test

**Review by gengjun-git — DISMISSED** [review]

**Comment by Astralidea** [comment]
run starrocks_fe_unittest

File: fe/fe-core/src/main/java/com/starrocks/catalog/ArrayType.java [hunk 3]
```diff
@@ -103,6 +103,18 @@ public ArrayType clone() {
         return clone;
     }
 
+    /**
+     * @return 33 (utf8_general_ci) if type is char varchar hll or bitmap
+     * 63 (binary) others
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/catalog/ArrayType.java:108]: please update the comment

**Comment by Astralidea** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 3 / 3 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/ArrayType.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/Type.java|2|2|100.00%|[]|

**Comment by Astralidea** [comment]
run starrocks_admit_test

**Review by kangkaisen — APPROVED** [review]

**Review by imay — APPROVED** [review]

**Comment by Astralidea** [comment]
@mergify backport branch-2.3

**Comment by Astralidea** [comment]
@mergify backport branch-2.2

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 3 / 3 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/ArrayType.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/Type.java|2|2|100.00%|[]|

**Pull request merged**